test/hotspot/jtreg/compiler/stringopts/TestFluidAndNonFluid.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2024 Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+ /*
+ * @test
+ * @bug 8341696
+ * @requires vm.compiler2.enabled
+ * @summary Allow C2 to also optimize non-fluid string builder calls.
+ * @library /test/lib /
+ * @run driver compiler.c2.irTests.stringopts.FluidAndNonFluid
+ */
+package compiler.c2.irTests.stringopts;
+
+import compiler.lib.ir_framework.*;
+import jdk.test.lib.Asserts;
+
+public class FluidAndNonFluid {
+
+    public static int unknown = 1;
+
+    public static void main(String[] args) {
+        // Dont inline any StringBuilder methods for this IR test to check if string opts are applied or not.
+        TestFramework.runWithFlags("-XX:CompileCommand=dontinline,java.lang.StringBuilder::*");
+    }
+
+    @DontInline
+    public static void opaque(StringBuilder builder) {
+        builder.append("Z");
+    }
+
+    @Run(test = {"fluid", "nonFluid", "nonFinal", "nonFluidExtraneousVariable", "nonFluidConditional",
+        "nonFluidOpaqueCall"})
+    public void runMethod() {
+        Asserts.assertEQ("0ac", fluidNoParam());
+        Asserts.assertEQ("ac", nonFluidNoParam());
+        Asserts.assertEQ("ac", fluid("c"));
+        Asserts.assertEQ("ac", nonFluid("c"));
+        Asserts.assertEQ("ac", nonFinal("c"));
+        Asserts.assertEQ("ac", nonFluidExtraneousVariable("c"));
+        Asserts.assertEQ("ac", nonFluidConditional("c"));
+        Asserts.assertEQ("aZ", nonFluidOpaqueCall());
+    }
+
+    @Test
+    @IR(failOn = {IRNode.ALLOC_OF, "StringBuilder", IRNode.CALL_OF_METHOD, "toString", IRNode.INTRINSIC_TRAP})
+    public static String fluidNoParam() {
+        return new StringBuilder("0").append("a").append("c").toString();
+    }
```

Comment:
Drive by comment: you seem to only have negative `failOn` tests here. You could consider adding a test where you have a positive rule, just to make sure you are matching the correct IR nodes.

Reply:
Never mind, I did not see the positive cases below :see_no_evil: